src/core/plugin/PluginController.cpp

```diff
@@ -1,54 +1,106 @@
-            auto plugin = std::make_unique<Plugin>(control, pluginPath.filename().string(), pluginPath);
-            if (!plugin->isValid()) {
-                g_warning("Error loading plugin \"%s\"", f.path().string().c_str());
-                continue;
+            try {
+                auto plugin = std::make_unique<Plugin>(control, pluginPath.filename().string(), pluginPath);
+                if (!plugin || !plugin->isValid()) {
+                    g_warning("Error loading plugin \"%s\"", f.path().string().c_str());
+                    continue;
+                }
+                plugin->setEnabled(plugin->isDefaultEnabled());
+                returner.emplace_back(std::move(plugin));
+            } catch (std::exception const& e) {
+                g_warning("Error loading plugin \"%s\": %s", f.path().string().c_str(), e.what());
             }
+        }
+    } catch (fs::filesystem_error const& e) {
+        g_warning("Could not open plugin dir: \"%s\": %s", path.string().c_str(), e.what());
+    }
+    return returner;
+}
 
-            if (plugin->isDefaultEnabled()) {
-                plugin->setEnabled(!(std::find(pluginDisabled.begin(), pluginDisabled.end(), plugin->getName()) !=
-                                     pluginDisabled.end()));
-            } else {
-                plugin->setEnabled(std::find(pluginEnabled.begin(), pluginEnabled.end(), plugin->getName()) !=
-                                   pluginEnabled.end());
-            }
+auto emplace_sorted_if_not_exists(std::vector<std::unique_ptr<Plugin>>& plugins, std::unique_ptr<Plugin> plugin)
+        -> void {
+    auto const& path = plugin->getPath();
+    auto endit = std::end(plugins);
+    auto it = std::find_if(std::begin(plugins), endit,
+                           [&path](auto const& p) { return fs::canonical(p->getPath()) == fs::canonical(path); });
+    if (it != endit) {
+        return;
+    }
+    plugins.emplace_back(std::move(plugin));
+    std::sort(begin(plugins), end(plugins), [](auto const& lhs, auto const& rhs) {
+        auto tieer = [](auto&& p) { return std::tie(p->getName(), p->getPath()); };
+        return tieer(lhs) < tieer(rhs);
+    });
+}
+}  // namespace
+#endif
 
-            plugin->loadScript();
 
-            this->plugins.emplace_back(std::move(plugin));
+PluginController::PluginController(Control* control): control(control) {
+#ifdef ENABLE_PLUGINS
+    // Todo(fabian) move those search paths into PathUtils
+    auto searchPath = control->getGladeSearchPath()->getFirstSearchPath();
+    auto searchPaths = {fs::weakly_canonical(searchPath /= "../plugins"),  //
+                        fs::weakly_canonical(searchPath /= "/plugins"),    //
+                        Util::getConfigSubfolder("plugins")};
+
+    for (auto&& path: searchPaths) {
+        for (auto&& plugin: load_awailable_plugins_from(path, control)) {
+            emplace_sorted_if_not_exists(plugins, std::move(plugin));
+        }
```

Comment:
This is O(n^2). Could we improve this by using an ordered map (indexed by path of plugin folder) to track the loaded plugins?

Reply:
Yes this can be improved, for example with an boost::sorted_vector(?) Or by using std::lower_bound.

I decided to use a vector over a map, because we first won't have many Plugins so that it matters. 
Second, a std::map operation might be 100 times slower.
Third the effort to implement my own container is not worth it (using boost is not an option currently).

Ps until yesterday, I used a std::map for this.

Reply:
This is kind of a small perf improvement thing but also a low hanging fruit. I think the easiest and most effective improvement for now is to use `std::lower_bound` (as long as we ensure that plugins is pre-sorted, which should not be difficult to implement and maintain). For the low plugin count case, the overhead is negligible; and this should hopefully scale better in the high plugin count case.